### ISSUE TITLE
tag/value "freeform" search for advanced users?

### ISSUE THREAD

**Issue by RicoZ-OSM** [osmandapp/Osmand#2209]
Is it possible to search for specific tags or tag/values and combinations thereof by entering a search query similar to how it can be done in JOSM? Or could it be implemented, at least for tags known in poi_types.xml ?

**Comment by vshcherb** [comment]
Isn't it already there? You could see suggestions popup when you type.

**Comment by RicoZ-OSM** [comment]
not quite, or I haven't found how :
* can't search for poi_additional directly
* can't do "complex" queries

**Comment by RicoZ-OSM** [comment]
looking at poi_types.xml, the distinction between poi_type and poi_additional makes the stored data hard to search or next to impossible to find in some cases:

* no easy way to search medical specialities like places doing cesarean section, ophtalmotology etc?
* you can search for car_wash as POI but not washing_machine which is poi_additional?
* you can search for sport=swimming but not architecture_gothic, plum trees,  banana trees or places marked with nudism
* when stuck in Sidney/Montana at 23:05 I  might want to search for "anything open at this time"

There is an incredible wealth of information in the maps which is hard to retrieve.

**Comment by vshcherb** [comment]
poi additional search is unlikely to be implemented because of how the data stored. So it is not really related to the original issue itself. In case you know exact poi type it is related to, search still could be implemented similar to fuel:diesel and so on. 
I think the issue might split into 2 parts, change data format in order to make additional searchable and change ui or create additional filters in order to make easy to find particular poi's

**Comment by RicoZ-OSM** [comment]
Imo the distinction between poi and poi_additional is very soft, rather artificial. You can search bridge:structure=humpback (that is nice) but can't find architecture=gothic although by conception those are pretty similar.

So I would advocate storing both types of points equal and have poi_additional defined only in the poi*xml file as a search/filter/ordering/category hint. Unless of course, it would have a large negative effect on map size.

Another approach would be to support searches like  poi=* &&  poi_additional=query . Would take longer to search but might still work good enough for many uses.

As of the ui it might be an idea to implement query strings (similar to JOSM) first and design the ui once people have played with that a bit.

**Comment by vshcherb** [comment]
It is still subject to discuss. Let's take, 5-stars hotels are additional to hotels and they should be searched inside of the hotel poi type and it is possible to add them today, the same as we did fuel types which are additional. Of course there are other 5-star objects but 2 star restaurant will mean completely different then 2 star hotel. 
Sometimes additional objects has connection but still may not be that strict, like religion relates to church and to cemetery. 

Poi additional are designed as UI filter on top of the category. Very typical examples are open, cuisine. We can't enumerate them, but we can filter. And we can't implement in JOSM because OsmAnd doesn't store all original tag/values, so the service like overpass simply can't work in OsmAnd.

I think to close that issue cause it is misleading the exact direction we could take. I would propose 2 ways especially related to poi additional, first of all make sure that data is present in obf for specific pois and make sure that filter UI (when you select poi category and click icon filter on the bottom) is handling additionals

**Comment by RicoZ-OSM** [comment]
I think the concept of poi & poi_additional as a strict hierarchy is unnecessary restrictive.

I would like to view it as a kind of "soft" hierarchy. Most of the time it makes sense to search 5stars as one of either hotel or restaurant - but if someone wants to search all 5 stars objects somewhere around and the data is available I would not forbid it.

If someone starts searching hotels he could get a suggestion to filter by stars.
If someone starts searching by stars he should get a suggestion to filter by hotels and/or restaurants, campgrounds...

I someone starts searching by opening times don't worry to give filter suggestions:)

As of hierarchy consider that an ATM could be considered belonging to either of commerce, shopping, tourism accommodation or dining parent objects.

In summary, I would
* treat poi and poi_additional equal
* instead of a hierarchy use "likely associations" (2way) to suggest additional filters
* there could be some value 1-5 expressing how much a poi type is expected to be used/useful as "main search" or rather as additional filter which would influence when/how the particular poi-type would be suggested

**Comment by vshcherb** [comment]
Poi type/additional structure was implemented mostly to create proper UI and to help translating stuff. 
Hierarchy is needed to create custom filters where you have poi category -> poi type. Everything that's not poi type is poi additional. Hierarchy also helps to store data efficiently, though in small cases it creates troubles. Like POI has some characteristics which could be searched independently (bank has atm). There is no issue to search all POIs and filter open POIs, there was a real request for this and we implemented it, the only thing it could much slower than per given type which is explainable. 

Structure is implemented in such a way that it reflects UI. In case  we want to have different UI it is explainable. What I read from that issue that there are users who know internals of OSM and they would like to search by tags (similar to overpass which is very flexible). Most of the users don't know that stuff at all and the behavior is to search hotel and check each of them if it is 5 stars. Or search hotels and click on filter for hotel. 

We didn't implement that part of UI though we hope to do next release. It will look like this.
- You type "Hotel" - press on the suggestion
- Screen shows up and in the top you will see buttons "5 star hotel", ... "Open hotels", ... or filter by name hotel 
- Then under the buttons there is a full list of hotels around you.

It already works like that but poi additional is not intuitive to find. It doesn't give full flexibility but it follows standard person behavior.

For OSM users IMHO should be completely different search.
